shared/modules/metametrics.js

```diff
@@ -0,0 +1,196 @@
+ * @property {string} network - the name of the current network
+ * @property {string} [metaMetricsSendCount] - number of transactions sent, used to add metametricsId
+ *                                        - intermittently to events with onchain data attached to them
+ *                                        - used to protect identity of users.
+ */
+
+/**
+ * @typedef {Object} MetaMetricsEventPayload
+ * @property {string}  event - event name to track
+ * @property {string}  category - category to associate event to
+ * @property {boolean} [isOptIn] - happened during opt in/out workflow
+ * @property {object}  [properties] - object of custom values to track, snake_case
+ * @property {number}  [revenue] - amount of currency that event creates in revenue for MetaMask
+ * @property {string}  [currency] - ISO 4127 format currency for events with revenue, defaults to US dollars
+ * @property {number}  [value] - Abstract "value" that this event has for MetaMask.
+ * @property {boolean} [excludeMetaMetricsId] - whether to exclude the user's metametrics id for anonymity
+ * @property {MetaMetricsEventContext} [additionalContext] - additional context to attach to event
+ */
+
+/**
+ * Returns a function for tracking Segment events.
+ *
+ * @param {string} metamaskVersion - The current version of the MetaMask extension.
+ * @param {() => MetaMetricsRequiredState} getDynamicState - A function returning required fields
+ * @param {boolean} defaultMetaMetricsExclusion - Used to set the default state of exludeMetaMetricsId
+ * @returns {(payload: MetaMetricsEventPayload) => Promise<never>} - function to track an event
+ */
+export function getTrackMetaMetricsEvent (
+  metamaskVersion,
+  getDynamicState,
+  defaultMetaMetricsExclusion = false,
+) {
+  const version = process.env.METAMASK_ENVIRONMENT === 'production'
+    ? metamaskVersion
+    : `${metamaskVersion}-${process.env.METAMASK_ENVIRONMENT}`
+
+  return function trackSegmentEvent ({
+    event,
+    category,
+    isOptIn,
+    properties = {},
+    revenue,
+    currency,
+    value,
+    excludeMetaMetricsId: excludeId,
+    additionalContext = {},
+  }) {
+    if (!event || !category) {
+      throw new Error('Must specify event and category.')
+    }
+    const {
+      participateInMetaMetrics,
+      context: providedContext,
+      metaMetricsId,
+      environmentType,
+      chainId,
+      network,
+      metaMetricsSendCount,
+    } = getDynamicState()
+
+    let excludeMetaMetricsId = excludeId ?? defaultMetaMetricsExclusion
+
+    // This is carried over from the old implementation, and will likely need
+    // to be updated to work with the new tracking plan. I think we should use
+    // a config setting for this instead of trying to match the event name
+    const isSendFlow = Boolean(event.match(/^send|^confirm/u))
+    if (isSendFlow && metaMetricsSendCount && !sendCountIsTrackable(metaMetricsSendCount + 1)) {
```

Comment:
Looking at the [`ConfirmTransactionBase` component](https://github.com/MetaMask/metamask-extension/blob/develop/ui/app/pages/confirm-transaction-base/confirm-transaction-base.component.js#L475-L488), aren't we racing to increment the `metaMetricsSendCount`?

This may be a preexisting problem, but we should increment and update `metaMetricsSendCount` in a single place, ideally in some MetaMetrics handler function.

It's OK if we do that in a separate PR, but we should absolutely do it.

Reply:
I'll address this in a separate PR